### DISCUSSION TITLE
"No workspaces found" with incorrect npm workspace directory name when creating PR for open-telemetry/opentelemetry-js-contrib

### DISCUSSION THREAD

**Discussion by trentm** [renovatebot/renovate#30466] [Request Help] [answered]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Mend Renovate hosted app on github.com

### Please tell us more about your question or problem

Hello,
I am a maintainer of https://github.com/open-telemetry/opentelemetry-js-contrib. Recently the renovate job for that repository's `all patch versions` package rule started failing with this error message (this is an npm update):

```
⚠ Artifact update problem
Renovate failed to update an artifact related to this branch.
...
File name: package-lock.json
...
npm ERR! No workspaces found:
npm ERR!   --workspace=plugins/node/instrumentation-socketio
...
```

In the repo, the actual directory (and npm workspace) name is different: `plugins/node/instrumentation-socket.io` with a period:
https://github.com/open-telemetry/opentelemetry-js-contrib/tree/c1d67f8befb716c57bb5083a9d24f154128963df/plugins/node/instrumentation-socket.io

Is it possible that renovatebot is incorrectly normalizing the npm workspace dirs and removing that period in the workspace name?

FWIW, this *used* to work on this same dir. Here is an earlier renovate PR for the same `packageRules` entry that included an update to the `plugins/node/instrumentation-socket.io/package.json` file: https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1707

---

- comment on PR showing the update problem: https://github.com/open-telemetry/opentelemetry-js-contrib/pull/2006#issuecomment-1997939902
- renovate config: https://github.com/open-telemetry/opentelemetry-js-contrib/blob/c1d67f8befb716c57bb5083a9d24f154128963df/renovate.json#L6-L12

Thanks.

### Logs (if relevant)

Renovate job log that shows the failure: https://developer.mend.io/github/open-telemetry/opentelemetry-js-contrib/-/job/0190ffa6-a610-7f70-9100-66fbe2224894

<details><summary>Logs</summary>

```
...
DEBUG: Executing command (branch="renovate/all-patch")
{
  "command": "npm install --package-lock-only --no-audit --ignore-scripts --workspace=plugins/node/instrumentation-socketio socket.io@4.7.5 socket.io-client@4.7.5"
}

DEBUG: rawExec err (branch="renovate/all-patch")
{
  "err": {
    "cmd": "/bin/sh -c npm install --package-lock-only --no-audit --ignore-scripts --workspace=plugins/node/instrumentation-socketio socket.io@4.7.5 socket.io-client@4.7.5",
    "stderr": "npm ERR! No workspaces found:\nnpm ERR!   --workspace=plugins/node/instrumentation-socketio\n\nnpm ERR! A complete log of this run can be found in:\nnpm ERR!     /tmp/renovate/cache/others/npm/_logs/2024-07-29T18_39_18_175Z-debug-0.log\n",
    "stdout": "",
...
```

</details>

**Comment by trentm** [comment]
I did some local debugging of renovate for this (using https://github.com/trentm/renovatetestrepo1 as the test repo), and I believe this line has a bug:

https://github.com/renovatebot/renovate/blob/286036d4f3aad88bdd298fe8194a3b4b49f83c6e/lib/modules/manager/npm/post-update/npm.ts#L294

```js
     const workspaceDir = trimSlashes(packageFileDir.replace(lockFileDir, ''));
```

When debugging this for a run against my repo:
- `packageFileDir` is `plugins/node/instrumentation-socket.io`
- but `workspaceDir` is `plugins/node/instrumentation-socketio` (the `.` has incorrectly been removed)
- `lockFileDir` is `.`, so that explains why `packageFileDir.replace(lockFileDir, '')` is incorrectly removing the period

Perhaps that section could be changed to:

```diff
diff --git a/lib/modules/manager/npm/post-update/npm.ts b/lib/modules/manager/npm/post-update/npm.ts
index f00351061..b7215af48 100644
--- a/lib/modules/manager/npm/post-update/npm.ts
+++ b/lib/modules/manager/npm/post-update/npm.ts
@@ -291,7 +291,11 @@ export function divideWorkspaceAndRootDeps(
       );

       // workspaceDir = packageFileDir - lockFileDir
-      const workspaceDir = trimSlashes(packageFileDir.replace(lockFileDir, ''));
+      const workspaceDir = trimSlashes(
+        packageFileDir.startsWith(lockFileDir)
+          ? packageFileDir.slice(lockFileDir.length)
+          : packageFileDir
+      );

       if (is.nonEmptyString(workspaceDir)) {
         let workspaceName: string | undefined;
```

?

**Reply by trentm** [reply]
With that patch, renovate was able to create the PR successfully: https://github.com/trentm/renovatetestrepo1/pull/5

**Reply by trentm** [reply]
Should I submit a PR?
I don't know how to add a good test case for this because I don't know the repo's test suite at all.

**Reply by rarkins** [reply]
Please go ahead with the PR! We can help with testing

**Reply by trentm** [reply, marked as answer]
PR added: https://github.com/renovatebot/renovate/pull/30483
